api/src/main/java/jakarta/persistence/criteria/Expression.java

```diff
@@ -138,4 +139,126 @@ public interface Expression<T> extends Selection<T> {
+     * is equal to the argument, returning null if they are
+     * and the value of the first expression if they are not.
+     * @param y expression
+     * @return nullif expression
+     * @since 3.2
+     */
+    Expression<T> nullif(Expression<? extends T> y);
+
+    /**
+     * Create an expression that tests whether this expression
+     * is equal to the argument, returning null if they are
+     * and the value of the first expression if they are not.
+     * @param y value
+     * @return nullif expression
+     * @since 3.2
+     */
+    Expression<T> nullif(T y);
+
+    //case builders:
+
+    /**
+     * Create a simple case expression to test against this expression.
+     * @return simple case expression
+     * @since 3.2
+     */
+    <R> CriteriaBuilder.SimpleCase<T,R> selectCase();
+
+    //collection operations:
+
+    /**
+     * Create a predicate that tests whether this expression is
+     * a member of a collection.
+     * If the collection is empty, the predicate will be false.
+     * @param collection expression
+     * @return is-member predicate
+     * @since 3.2
+     */
+    <C extends Collection<T>> Predicate isMember(Expression<C> collection);
+
+    /**
+     * Create a predicate that tests whether this expression is
+     * not a member of a collection.
+     * If the collection is empty, the predicate will be true.
+     * @param collection expression
+     * @return is-not-member predicate
+     * @since 3.2
+     */
+    <C extends Collection<T>> Predicate isNotMember(Expression<C> collection);
+
+    //aggregate functions:
+
+    /**
+     * Create an aggregate expression applying the count operation.
+     * @return count expression
+     * @since 3.2
+     */
+    NumberExpression<Long> count();
+
+    /**
+     * Create an aggregate expression applying the count distinct
+     * operation.
+     * @return count distinct expression
+     * @since 3.2
+     */
+    NumberExpression<Long> countDistinct();
+
+    //ordering:
+
+    /**
+     * Create an ordering by the ascending value of this expression.
+     * @return ascending ordering corresponding to this expression
+     * @since 3.2
+     */
+    Order asc();
+
+    /**
+     * Create an ordering by the descending value of this expression.
+     * @return descending ordering corresponding to this expression
+     * @since 3.2
+     */
+    Order desc();
+
+    /**
+     * Create an ordering by the ascending value of this expression.
+     * @param nullPrecedence  the precedence of null values
+     * @return ascending ordering corresponding to this expression
+     * @since 3.2
+     */
+    Order asc(Nulls nullPrecedence);
+
+    /**
+     * Create an ordering by the descending value of this expression.
+     * @param nullPrecedence  the precedence of null values
+     * @return descending ordering corresponding to this expression
+     * @since 3.2
+     */
+    Order desc(Nulls nullPrecedence);
```

Comment:
Oh I see, they are declared here. But I wonder if it's really correct to have them here instead of on `ComparableExpression`. Yes, I know `CriteriaBuilder` allows them for any `Expression`, but maybe that's just a mistake?

I mean, if we're going to have `ComparableExpression` at all, it should be the interface defining things related to comparisons, it seems to me.

Reply:
Hard to say. I can tell you that e.g. `byte[]` or `char[]` are not `Comparable`, but on the database side, the type is most probably `varbinary` or `varchar` respectively, which both are orderable. Also, think about how converted types might not be `Comparable`, but the conversion target type is.

I understand the desire to make things "typesafe", but I honestly doubt that moving `asc`/`desc` to `ComparableExpression` would help. On the contrary, I think it might cause more trouble than good.

Reply:
I mean, it seems to me that sorting binary data is never meaningful. As for `char[]`, well, if the user had intended this array of characters to be interpreted as a string, then _surely_ they would have used `String` for that. So I think it's totally legit to say that these types can't be ordered. 

(Honestly, I don't even know why we even allow `char[]` except for "we allow it because we can".)

>  I honestly doubt that moving `asc`/`desc` to `ComparableExpression` would help.

If that's the case then I guess it's pretty unclear what purpose `ComparableExpression` even serves. If it's not abstracting over types which have a well-defined order, then it's not doing any work.

Reply:
`ComparableExpression` is abstracting over the relational predicates i.e. greater-than, smaller-than, etc.

I understand the desire to also move the ordering aspect into that, though like I was trying to say, I believe there are real use cases that might not work out this way. I'm fine with moving the methods if you insist and @lukasj agrees. I was just trying to point out why the existing `asc`/`desc` methods in `CriteriaBuilder` might be the way they are.